subsys/bluetooth/host/conn.c

```diff
@@ -747,11 +747,10 @@ static void conn_cleanup(struct bt_conn *conn)
 
 static void conn_destroy(struct bt_conn *conn, void *data)
 {
-	k_work_cancel_delayable(&conn->deferred_work);
-
-	conn->pending_no_cb = 0;
-
-	conn_cleanup(conn);
+	if (conn->state == BT_CONN_CONNECTED ||
+	    conn->state == BT_CONN_DISCONNECTING) {
```

Comment:
we don't need to handle the `connecting` states?

Reply:
As `bt_conn_set_state` has handling for the state change `BT_CONN_CONNECTING->BT_CONN_DISCONNECTED` I thought this was a safe operation.

Reply:
Hmm, looking at `bt_conn_set_state`, is there actually any reason why we don't unconditionally call `bt_conn_set_state(BT_CONN_DISCONNECT_COMPLETE)`?

Reply:
As there is specific handling for changing state `BT_CONN_CONNECTING/BT_CONN_CONNECTING_ADV` etc to `BT_CONN_DISONNECTED`, whereas `BT_CONN_CONNECTED/BT_CONN_DISCONNECTING` are designed to go through `BT_CONN_DISCONNECT_COMPLETE`